java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/client/ArrowFlightSqlClientHandler.java

```diff
@@ -92,11 +103,36 @@ public static ArrowFlightSqlClientHandler createNewHandler(final FlightClient cl
    * @param flightInfo The {@link FlightInfo} instance from which to fetch results.
    * @return a {@code FlightStream} of results.
    */
-  public List<FlightStream> getStreams(final FlightInfo flightInfo) {
-    return flightInfo.getEndpoints().stream()
-        .map(FlightEndpoint::getTicket)
-        .map(ticket -> sqlClient.getStream(ticket, getOptions()))
-        .collect(Collectors.toList());
+  public List<FlightStream> getStreams(final FlightInfo flightInfo) throws SQLException {
+    final List<FlightStream> allStreams = new ArrayList<>();
+
+    for (final FlightEndpoint endpoint : flightInfo.getEndpoints()) {
+      List<Location> locations = endpoint.getLocations();
+      if (locations.isEmpty()) {
+        allStreams.add(sqlClient.getStream(endpoint.getTicket(), getOptions()));
+        continue;
+      }
+      final Location location = locations.get(0); // purposefully discard other locations
+
+      logger.info(String.format("Getting a client for location %s.", location));
+      FlightSqlClient flightSqlClient;
+      try {
+        flightSqlClient = flightSqlClientPool.borrowObject(location);
+      } catch (final NoSuchElementException e) {
+        try {
+          flightSqlClientPool.addObject(location);
+          flightSqlClient = flightSqlClientPool.borrowObject(location);
+        } catch (final Exception addObjectEx) {
+          throw new SQLException("Failed to create and get a new FlightSqlClient in the ObjectPool.", addObjectEx);
+        }
+      } catch (final Exception borrowObjectEx) {
+        throw new SQLException("Failed to borrow a FlightSqlClient from the ObjectPool", borrowObjectEx);
+      }
+
+      allStreams.add(flightSqlClient.getStream(endpoint.getTicket(), getOptions()));
+      flightSqlClientPool.returnObject(location, flightSqlClient);
```

Comment:
We cannot return the client to the pool until we are done getting the stream. Otherwise it allows another query being executed to re-use the client while it's still being used.

Reply:
Maybe we can write an auto-closeable wrapper on top of FlightStream that returns to the pool on closure.